waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js

```diff
@@ -56,36 +27,5 @@ const registerPasswordHashing = (prismaClient) => {
 }
 
 export const registerAuthMiddleware = (prismaClient) => {
-  // NOTE: registerUserEntityValidation must come before registerPasswordHashing.
-  registerUserEntityValidation(prismaClient)
   registerPasswordHashing(prismaClient)
 }
-
-const userValidations = []
```

Comment:
Removing username & password validation from Prisma middleware

Reply:
I guess password hashing is fine there!

What is our reasoning though about removing username & pass validation from there -> aren't we loosing something? I know you said we can reintroduce it back later once we do refactoring, but why not fix it now in such way that we keep having it there? Do you think it is not useful to have it there? Should we discuss the use cases a bit, pros and cons of having or not having it there?

Reply:
Just to be clear: we are not removing the validation but rather moving it into the route 🙂

We are moving to route to make it work like email validation: connected to the login and signup process then connected to the Prisma model.

We are losing the customisation mechanism, but we should [reintroduce it ](https://github.com/wasp-lang/wasp/issues/1571) in a more user friendly way and in a way that enable also email validation customisation. 

The current customisation mechanism felt a bit hacky and it you needed to specify the overrides on each usage with `_waspCustomValidations`. Also, I assume the feature is not really used. I got the feeling that people like the higher level API much more.

Reply:
Sure that all makes sense, why I am asking is that that way it was before validation was enforced on any change to username / pass, not just the change via the routes. So if I as a developer do something more manually and want to set username/pass directly on the User entity, that would also be caught, while now it will not be caught. So I was curious why have you decided to drop that. For example, if I use https://wasp-lang.dev/docs/auth/username-and-pass#2-creating-your-custom-actions, I won't be getting these validations any more, right?

You say you think it wasn't used much -> I think there is certainly a case where it happens, which is https://wasp-lang.dev/docs/auth/username-and-pass#2-creating-your-custom-actions , and we since we are supporting that, I think we want to ensure good experience here. The question is, is it weird / bad if we drop these validations here? Is it expected by the developer? I think it might not be expected, so in that case we should mention this in the docs -> hey, you now have to do validation here on your own.
I wonder if they would rather have these validations here or not. If they are here, they can be annoying if they can't override them. But if there are none, they will have to implement their own.

I think I find it easier to think about if the concept of validating username/email is centralized per whole Wasp app, and you can either use it as it is default or customize it. I don't like the idea of them by accident not using it, or by accident going around it.

But I would like to hear your thoughts here, in more details!

Reply:
> For example, if I use https://wasp-lang.dev/docs/auth/username-and-pass#2-creating-your-custom-actions, I won't be getting these validations any more, right?

Yes, these validation won't apply here any more. The user will be responsible for validating the payload. And he can use our `ensureValidUsername` and `ensureValidPassword` helper functions. This is something I can write about in the docs in the section you linked (which I overlooked, sorry about that)

> You say you think it wasn't used much -> I think there is certainly a case where it happens, which is https://wasp-lang.dev/docs/auth/username-and-pass#2-creating-your-custom-actions

"a case where it happens" != "users using the feature", I was talking about users not really customising the lower level validation

> and we since we are supporting that, I think we want to ensure good experience here.

I agree we should give them tools to validate auth related fields or any fields in general 👍 

> I think I find it easier to think about if the concept of validating username/email is centralized per whole Wasp app, and you can either use it as it is default or customize it. 

I feel the validations are an auth feature and it should happen in that flow. If the users are building something of their own, they can use our lower level validation helpers like mentioned above.

Alternatively, validation is transparent to the user and it happens at the Prisma level which is a bit too hidden for my taste. I'd rather instruct them to use `ensureValidUsername` then explain the details of validation happening in the Prisma middleware. 

> I don't like the idea of them by accident not using it, or by accident going around it.

Users can't accidentally write their own login and signup actions. At that point they are responsible for validating the frontend payload before saving it in the database IMHO

Reply:
Ok @infomiho this makes sense to me, let's do this as you described here!